### PR TITLE
Refactor `Container` class and children

### PR DESCRIPTION
**Pull request by jcollins-g**
Rename and change semantics for many getters on the `Container` class.

This makes Dartdoc's use of terminology for class members more closely aligned with Dart and Analyzer internals, and improves separation between the generation layer and Dartdoc internals.

- "Properties" are now called `Fields` internally outside of the templates themselves.
- Move everything having to do with inheritance to the `Class` class out of `Container`.
- Move everything common to all `Container`s out of `Class` and into `Container` and eliminate some duplication.
- Use a consistent naming scheme for template getters in particular:
  - `sorted` for getters where ordering matters.  Don't store and sort lists of items over and over for non-documented ModelElements.
  - `staticFields` now includes constant static fields, `variableStaticFields` has the old behavior.
  - `instanceMethods` and `instanceFields` for non-statics, rather than `methods` excluding statics by default.
- Compose getters out of words with consistent meanings, e.g. `hasPublicInheritedFields` and `publicConstantFields` rather than using so many abbreviations, so templates are clearer.

Note:  this completely breaks markdown generation until the templates are updated.  Will add a smoke test for markdown and update those templates in a followup.

### CODE REVIEW

**Comment by jcollins-g** [comment]
I didn't want to hand-edit all the templates so I just copied the whole directory across for the templates-dir test, modifying only the index so it's possible to tell the difference by examining the output (and therefore verifying that dartdoc is using the right template dir).

There is probably a better way to do this, open to better ideas.

File: lib/src/model/class.dart [hunk 1]
```diff
@@ -532,22 +429,17 @@ class Class extends Container
...
-    _methods ??= element.methods.map((e) {
-      return ModelElement.from(e, library, packageGraph) as Method;
-    }).toList(growable: false)
-      ..sort(byName);
-    return _methods;
-  }
+  Iterable<Method> get declaredMethods =>
+      _declaredMethods ??= element.methods.map((e) {
+        return ModelElement.from(e, library, packageGraph) as Method;
+      }).toList(growable: false);
```
Comment (srawlins) [inline lib/src/model/class.dart:440]: Hmm, does the `toList` here add anything important?

File: lib/src/model/container.dart [hunk 2]
```diff
@@ -5,121 +5,215 @@
...
   Container(Element element, Library library, PackageGraph packageGraph)
       : super(element, library, packageGraph, null);
 
   bool get isClass => element is ClassElement;
-
   bool get isExtension => element is ExtensionElement;
 
-  List<Method> get methods => [];
+  @mustCallSuper
+  Iterable<ModelElement> get allModelElements => quiver.concat([
```
Comment (srawlins) [inline lib/src/model/container.dart:38]: Hmm, should we just use spread here? Does quiver.concat buy us something in a getter?

File: lib/src/model/container.dart [hunk 3]
```diff
@@ -5,121 +5,215 @@
...
+        staticAccessors,
+        staticMethods,
+      ]);
 
-  List<Method> get instanceMethods {
-    if (_instanceMethods != null) return _instanceMethods;
+  /// All methods, including operators and statics, declared as part of this
+  /// [Container].  [declaredMethods] must be the union of [instanceMethods],
+  /// [staticMethods], and [instanceOperators].
+  ///
```
Comment (srawlins) [inline lib/src/model/container.dart:51]: Blank line?

File: lib/src/model/container.dart [hunk 4]
```diff
@@ -5,121 +5,215 @@
...
+          _membersByName[me.name] = [];
+        }
+        _membersByName[me.name].add(me);
+      }
+    }
+    ModelElement member;
+    var possibleMembers = _membersByName[example.name]
+        .where((e) => e.runtimeType == example.runtimeType);
+    if (example.runtimeType == Accessor) {
+      possibleMembers = possibleMembers.where(
```
Comment (srawlins) [inline lib/src/model/container.dart:159]: Can this code be updated to use whereType? It seems very strange to use runtimeType here...

**Comment by jdkoren** [comment]
> I didn't want to hand-edit all the templates so I just copied the whole directory across for the templates-dir test, modifying only the index so it's possible to tell the difference by examining the output (and therefore verifying that dartdoc is using the right template dir).

Checking the index afterward for a specific string seems reasonable.

I sort of liked that the template directories were quite different as it ensures that we don't inadvertently rely on specific filenames or directory structure (aside from what's documented in the help string). Not sure how strongly I feel about it, though.

**Comment by jcollins-g** [comment]
@jdkoren I hear you on not wanting to throw out any coverage we might have been getting there.  However, code coverage results suggests we didn't lose much by changing to a simpler construction, though.  Yes, we probably should be testing edge cases in custom templating more effectively.  Maybe at some point we can do that with more deliberate intention.

**Pull request merged**